### PR TITLE
Add single definition of status codes

### PR DESCRIPTION
**Pull request by cmccoy**
Introduces `Status`, which defines canonical status results to return from operations.
Updates all clients to use these codes.

### CODE REVIEW

File: core/src/main/java/com/yahoo/ycsb/StatusCode.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  /**
+   * The operation failed, with a generic error.
+   */
+  public static final int ERROR = -1;
+
+  /**
+   * The operation failed: the requested record was not found.
+   */
+  public static final int NOT_FOUND = -3;
+}
```
Comment (kruthar) [inline core/src/main/java/com/yahoo/ycsb/StatusCode.java]: Some other error codes that could be useful here are:
* UNEXPECTED_RESULT - successful operation, but result was not what was expected (eg, wrong number of records, etc.)
* CLIENT_ERROR - YCSB had an error running an operation (ex. given tablename was null)
* UNIMPLEMENTED - used with operation is not implemented for the given module

I'm sure there are others that could be useful, but initially just creating this centralized class will allow for easy updating and canonical status codes.

**Comment by allanbank** [comment]
@cmccoy - I broke this with the merge of the checkstyle changes. Want me to take a swing at cleaning it up?

File: mongodb/src/main/java/com/yahoo/ycsb/db/AsyncMongoDbClient.java [hunk 2]
```diff
@@ -146,12 +147,12 @@ public final int delete(final String table, final String key) {
       final long res = collection.delete(q, writeConcern);
       if (res == 0) {
         System.err.println("Nothing deleted for key " + key);
-        return 1;
+        return StatusCode.ERROR;
```
Comment (allanbank) [inline mongodb/src/main/java/com/yahoo/ycsb/db/AsyncMongoDbClient.java]: This should be Status.NOT_FOUND.

File: mongodb/src/main/java/com/yahoo/ycsb/db/AsyncMongoDbClient.java [hunk 3]
```diff
@@ -360,10 +361,10 @@ public final int read(final String table, final String key,
       if (queryResult != null) {
         fillMap(result, queryResult);
       }
-      return queryResult != null ? 0 : 1;
+      return queryResult != null ? StatusCode.OK : StatusCode.ERROR;
```
Comment (allanbank) [inline mongodb/src/main/java/com/yahoo/ycsb/db/AsyncMongoDbClient.java]: StatusCode.ERROR should be Status.NOT_FOUND.

File: mongodb/src/main/java/com/yahoo/ycsb/db/AsyncMongoDbClient.java [hunk 4]
```diff
@@ -412,7 +413,7 @@ public final int scan(final String table, final String startkey,
       final MongoIterator<Document> cursor = collection.find(find);
       if (!cursor.hasNext()) {
         System.err.println("Nothing found in scan for key " + startkey);
-        return 1;
+        return StatusCode.ERROR;
```
Comment (allanbank) [inline mongodb/src/main/java/com/yahoo/ycsb/db/AsyncMongoDbClient.java]: NOT_FOUND again.

File: mongodb/src/main/java/com/yahoo/ycsb/db/AsyncMongoDbClient.java [hunk 5]
```diff
@@ -461,10 +462,10 @@ public final int update(final String table, final String key,
       }
       final long res =
           collection.update(query, update, false, false, writeConcern);
-      return res == 1 ? 0 : 1;
+      return res == 1 ? StatusCode.OK : StatusCode.ERROR;
```
Comment (allanbank) [inline mongodb/src/main/java/com/yahoo/ycsb/db/AsyncMongoDbClient.java]: ERROR --> NOT_FOUND.

File: mongodb/src/main/java/com/yahoo/ycsb/db/MongoDbClient.java [hunk 6]
```diff
@@ -152,12 +152,12 @@ public int delete(String table, String key) {
           collection.withWriteConcern(writeConcern).deleteOne(query);
       if (result.wasAcknowledged() && result.getDeletedCount() == 0) {
         System.err.println("Nothing deleted for key " + key);
-        return 1;
+        return StatusCode.ERROR;
```
Comment (allanbank) [inline mongodb/src/main/java/com/yahoo/ycsb/db/MongoDbClient.java]: ERROR->NOT_FOUND

File: mongodb/src/main/java/com/yahoo/ycsb/db/MongoDbClient.java [hunk 7]
```diff
@@ -333,10 +333,10 @@ public int read(String table, String key, Set<String> fields,
       if (queryResult != null) {
         fillMap(result, queryResult);
       }
-      return queryResult != null ? 0 : 1;
+      return queryResult != null ? StatusCode.OK : StatusCode.ERROR;
```
Comment (allanbank) [inline mongodb/src/main/java/com/yahoo/ycsb/db/MongoDbClient.java]: ERROR->NOT_FOUND.

File: mongodb/src/main/java/com/yahoo/ycsb/db/MongoDbClient.java [hunk 8]
```diff
@@ -384,7 +384,7 @@ public int scan(String table, String startkey, int recordcount,
 
       if (!cursor.hasNext()) {
         System.err.println("Nothing found in scan for key " + startkey);
-        return 1;
+        return StatusCode.ERROR;
```
Comment (allanbank) [inline mongodb/src/main/java/com/yahoo/ycsb/db/MongoDbClient.java]: ERROR->NOT_FOUND.

File: mongodb/src/main/java/com/yahoo/ycsb/db/MongoDbClient.java [hunk 9]
```diff
@@ -440,12 +440,12 @@ public int update(String table, String key,
       UpdateResult result = collection.updateOne(query, update);
       if (result.wasAcknowledged() && result.getMatchedCount() == 0) {
         System.err.println("Nothing updated for key " + key);
-        return 1;
+        return StatusCode.ERROR;
```
Comment (allanbank) [inline mongodb/src/main/java/com/yahoo/ycsb/db/MongoDbClient.java]: ERROR->NOT_FOUND

**Comment by cmccoy** [comment]
@allanbank - I'm happy to take a pass at the merge conflicts. Thanks for looking over the mongodb bindings.

File: core/src/main/java/com/yahoo/ycsb/StatusCode.java (hunk 1 above)
Comment (cmccoy) [inline core/src/main/java/com/yahoo/ycsb/StatusCode.java]: Added UNEXPECTED_RESULT and UNIMPLEMENTED.

CLIENT_ERROR sounds like a failed assertion - can we just throw in that case?

**Comment by cmccoy** [comment]
Actually, having integer status codes doesn't make a ton of sense to me. I replaced those with a simple Status class.

Bindings with more specific errors (e.g., timeouts) return additional statuses.

This changes the default output from messages like:

    [UPDATE], Return=0, 511

To:

    [UPDATE], Return=OK, 511

**Comment by kruthar** [comment]
Nice! I like that, consolidated and self documenting.

**Comment by cmccoy** [comment]
@kruthar: I also removed the table/key `null` checks from the JDBC binding - none of the other bindings have that check, I think throwing is more appropriate.

**Comment by cmccoy** [comment]
@allanbank: thanks for reviewing the MongoDB section - updated to NOT_FOUND where you specified.

**Comment by allanbank** [comment]
:+1:

@cmccoy - I will let you decide if you want this in 0.5.0 or not.

**Comment by cmccoy** [comment]
Thanks.

**Pull request merged**